File: python/benches/blast_shape.py
"""Four-shape body-pool builder for the Phase 19 Python bench harness.

Mirrors ``crates/beava-bench/src/blast_shape.rs`` semantics:

  - ``fixed`` — one body, reused N times (cache-warm marketing peak)
  - ``uniform`` — ``user_id`` rolls evenly over ``cardinality`` keys (cache-pessimistic)
  - ``zipfian`` — Zipfian distribution over rank with alpha=1.0 default (realistic fraud)
  - ``mixed`` — round-robin over ``mixed_event_count`` distinct event names

REVISED D-09 (commit 88f1161) per CONTEXT.md: the pool stores Python ``dict`` bodies
(plus an event-name string for ``mixed``), NOT pre-encoded frame bytes. Encoding
happens INSIDE ``transport.send_push()`` (TCP) or ``transport._client.post()`` (HTTP)
per call — this is the SDK overhead the bench is honestly measuring; matches what
a real ``app.push()`` user would see when SDK-APP-04 lands.

Raw ``socket.create_connection + sock.sendall(pre_encoded_bytes)`` is FORBIDDEN.
"""

from __future__ import annotations

import random
from dataclasses import dataclass, field
from typing import Any, Literal

BlastShapeName = Literal["fixed", "uniform", "zipfian", "mixed"]
WireFormatName = Literal["json", "msgpack"]
TransportName = Literal["http", "tcp"]


@dataclass
class PoolConfig:
    """Configuration for a per-worker pool of event bodies."""

    shape: BlastShapeName
    wire_format: WireFormatName
    transport: TransportName
    cardinality: int
    zipf_alpha: float
    mixed_event_count: int
    seed: int
    pipeline_event_name: str
    pipeline_key_field: str
    pipeline_extra_fields: dict[str, str]
    # For mixed shape: synthesized event-name list. If pipeline only registers
    # one event the harness falls back to suffixed clones of the primary name.
    mixed_event_names: list[str] = field(default_factory=list)


@dataclass
class PoolItem:
    """One event in the pool: which event name to push + the body dict."""

    event_name: str
    body: dict[str, Any]


def _build_body(
    cfg: PoolConfig,
    key_idx: int,
    seq: int,
    rng: random.Random,
) -> dict[str, Any]:
    """Build one event body with key + event_time + scalar extra fields."""
    body: dict[str, Any] = {
        cfg.pipeline_key_field: f"k{key_idx:08d}",
        "event_time": 1_000_000 + seq,
    }
    for fld, ty in cfg.pipeline_extra_fields.items():
        if ty == "f64":
            body[fld] = round(rng.uniform(0.0, 1000.0), 4)
        elif ty == "i64":
            body[fld] = rng.randint(0, 999_999)
        elif ty == "str":
            body[fld] = f"s{rng.randint(0, 999)}"
        else:
            # Unknown type: stash 0 so the server's schema check still passes for
            # numeric primitives. Anything truly exotic should be added explicitly.
            body[fld] = 0
    return body


def _zipf_zeta(n: int, alpha: float) -> float:
    """Zeta value used by the rank-based Zipfian sampler."""
    total: float = 0.0
    for i in range(1, n + 1):
        total += 1.0 / (i**alpha)
    return total


class _ZipfianSampler:
    """Rank-based Zipfian sampler.

    Mirrors the recipe in ``rand_distr::Zipf`` semantics for alpha != 1: classic
    Gray et al. rejection sampler over rank ``r in [0, k)`` with
    ``P(r) ∝ 1/(r+1)^alpha``. Pure Python; not tuned for absolute speed, but
    fast enough for pool sizes up to a few million.
    """

    def __init__(self, k: int, alpha: float, seed: int) -> None:
        if k <= 0:
            raise ValueError(f"k must be > 0, got {k}")
        if alpha <= 0:
            raise ValueError(f"alpha must be > 0, got {alpha}")
        if alpha == 1.0:
            # alpha=1 hits a singularity in the eta formula below. Use a slightly
            # offset alpha to keep the closed-form sampler well-defined; the
            # statistical difference at alpha=1.0 vs 1.0001 is negligible for
            # bench purposes and matches Gray et al.'s original published recipe.
            alpha = 1.0001
        self.k = k
        self.alpha = alpha
        self.rng = random.Random(seed)
        self.zetan = _zipf_zeta(k, alpha)
        zeta2 = _zipf_zeta(2, alpha)
        # eta as defined in Gray et al. for the rank-based rejection sampler.
        self.eta = (
            (1.0 - (2.0 / k) ** (1.0 - alpha))
            / (1.0 - zeta2 / self.zetan)
        )

    def sample(self) -> int:
        """Return a rank in ``[0, k)`` distributed Zipf(alpha)."""
        u = self.rng.random()
        uz = u * self.zetan
        if uz < 1.0:
            return 0
        if uz < 1.0 + 0.5**self.alpha:
            return 1
        v = self.rng.random()
        rank = int(
            self.k * ((self.eta * v - self.eta + 1.0) ** (1.0 / (1.0 - self.alpha)))
        )
        return min(max(rank, 0), self.k - 1)


def build_pool(cfg: PoolConfig, n: int) -> list[PoolItem]:
    """Build N pool items matching the requested shape.

    REVISED D-09: each item is ``PoolItem(event_name, body_dict)``. Encoding
    happens inside the transport.send_push / transport._client.post call site
    in blast.py. Raw frame bytes are NOT pre-built here.

    Args:
        cfg: PoolConfig with shape + wire_format + transport + RNG seed.
        n: Number of items to build.

    Returns:
        list[PoolItem] of length ``n``.
    """
    if n <= 0:
        return []

    rng = random.Random(cfg.seed)
    pool: list[PoolItem] = []

    if cfg.shape == "fixed":
        # ONE pre-built body, reused N times. Cache-warm marketing peak.
        body = _build_body(cfg, key_idx=0, seq=0, rng=rng)
        item = PoolItem(event_name=cfg.pipeline_event_name, body=body)
        return [item] * n

    if cfg.shape == "uniform":
        # user_id rolls evenly over `cardinality` keys.
        cardinality = max(cfg.cardinality, 1)
        for seq in range(n):
            key_idx = rng.randrange(0, cardinality)
            body = _build_body(cfg, key_idx, seq, rng)
            pool.append(PoolItem(event_name=cfg.pipeline_event_name, body=body))
        return pool

    if cfg.shape == "zipfian":
        cardinality = max(cfg.cardinality, 1)
        sampler = _ZipfianSampler(cardinality, cfg.zipf_alpha, cfg.seed ^ 0xDEAD)
        for seq in range(n):
            key_idx = sampler.sample()
            body = _build_body(cfg, key_idx, seq, rng)
            pool.append(PoolItem(event_name=cfg.pipeline_event_name, body=body))
        return pool

    if cfg.shape == "mixed":
        m = max(cfg.mixed_event_count, 1)
        if len(cfg.mixed_event_names) >= m:
            names = list(cfg.mixed_event_names[:m])
        else:
            base = list(cfg.mixed_event_names) if cfg.mixed_event_names else [
                cfg.pipeline_event_name
            ]
            # Pad with suffixed clones of the primary name. A single-event
            # config is the common case; the suffixed names will not exist on
            # the server, but the harness itself is exercised.
            while len(base) < m:
                base.append(f"{cfg.pipeline_event_name}_{len(base)}")
            names = base[:m]

        cardinality = max(cfg.cardinality, 1)
        for seq in range(n):
            ev = names[seq % m]
            key_idx = rng.randrange(0, cardinality)
            body = _build_body(cfg, key_idx, seq, rng)
            pool.append(PoolItem(event_name=ev, body=body))
        return pool

    raise ValueError(f"unknown shape: {cfg.shape!r}")

File: crates/beava-bench/src/blast_shape.rs
//! Pool=N pre-encoded TCP-frame builder for the standalone bench binaries.
//!
//! 1. **Why Pool=N (not a sampler):** pre-encoding ALL N frames at sender
//!    startup eliminates per-iteration RNG cost AND per-iteration encode
//!    cost from the bench hot loop. The bench-side floor becomes "as fast
//!    as TCP `write_all` can drain" — the server-side ceiling is the only
//!    number we're measuring. Pool memory is ~500 MB-1 GB for N = 1 M;
//!    callers are responsible for sizing against host RAM.
//! 2. **Setup time excluded from `wall_clock_ms`:** [`build_pool_timed`]
//!    returns `(Vec<Bytes>, Duration)` so the caller can subtract pool
//!    setup from any saturation measurement.
//! 3. **All four shapes share this abstraction:** fixed / uniform / zipfian
//!    / mixed are emitted in identical TCP-frame envelopes (`CT_JSON` or
//!    `CT_MSGPACK`) so the ledger rows are directly comparable.
//! 4. **Determinism:** every shape uses `StdRng::seed_from_u64(seed)` and
//!    Zipfian's hybrid sampler is seeded the same way; same-seed runs
//!    produce byte-identical pool output.
//!
//! The module reuses `beava_core::wire::encode_frame` verbatim — zero new
//! wire formats.

use beava_core::wire::{encode_frame, Frame, CT_JSON, CT_MSGPACK, OP_PUSH};
use bytes::{Bytes, BytesMut};
use rand::rngs::StdRng;
use rand::{Rng, SeedableRng};
use serde::Serialize;
use serde_json::Value;
use std::time::{Duration, Instant};

/// The per-pipeline configuration consumed by the pool builder.
///
/// Field layout matches the binary harness (`src/bin/beava-bench-v18.rs`) so
/// the JSON files under `crates/beava-bench/configs/` deserialize directly
/// into either type without conversion.
#[derive(Debug, Clone, serde::Deserialize)]
pub struct PipelineConfig {
    pub name: String,
    // reason: deserialized from configs/*.json; retained for serde-shape
    // parity with the binary harness even if the blast_shape path doesn't
    // read this field.
    #[allow(dead_code)]
    pub description: String,
    pub register: Value,
    pub event_name: String,
    pub features: Vec<String>,
    pub key_field: String,
    pub extra_fields: serde_json::Map<String, Value>,
}

/// Wire format for the encoded pool. Matches `bin/beava-bench-v18.rs`.
///
/// JSON is the curl/HTTP-default; MessagePack is the production fast-path on
/// TCP for blast benches.
#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum WireFormat {
    Json,
    Msgpack,
}

/// One of the four blast shapes — controls how the pool's per-frame
/// `key_field` and `event_name` are sampled.
#[derive(Copy, Clone, Debug)]
pub enum BlastShape {
    /// One pre-encoded frame, reused N times. Cache-warm marketing peak.
    Fixed,
    /// `key_field` rolls evenly over `cardinality` keys. Cache-pessimistic
    /// floor.
    Uniform { cardinality: u64 },
    /// Zipfian distribution over `cardinality` keys with `alpha` skew.
    /// Realistic fraud workload.
    Zipfian { alpha: f64, cardinality: u64 },
    /// Pool spans `event_count` registered events sampled per push.
    /// Multi-stream realism.
    Mixed { event_count: usize },
}

/// Static configuration passed alongside `BlastShape` into `build_pool`.
pub struct BlastShapeConfig<'a> {
    pub pipeline: &'a PipelineConfig,
    /// For `BlastShape::Mixed` only. Must contain at least `event_count`
    /// names. For other shapes, ignored.
    pub event_names_for_mixed: &'a [&'a str],
    pub wire_format: WireFormat,
    pub seed: u64,
}

/// Errors returned by `build_pool`. None of these are recoverable — the
/// caller is misconfigured.
#[derive(Debug, thiserror::Error)]
pub enum BlastShapeError {
    #[error("BlastShape::Mixed requires at least event_count distinct event names in event_names_for_mixed")]
    MixedRequiresMultipleEvents,
    #[error("Zipfian alpha must be > 0.0")]
    InvalidAlpha,
    #[error("cardinality must be > 0 (Uniform / Zipfian)")]
    InvalidCardinality,
}

/// Deterministic, seedable Zipfian rank sampler.
///
/// Implements the hybrid algorithm from Gray et al. ("Quickly Generating
/// Billion-Record Synthetic Databases", SIGMOD 1994). For `alpha = 1.0` the
/// limiting form `(self.eta * v - self.eta + 1.0).powf(1.0 / (1.0 - alpha))`
/// degenerates (division by zero); we route the alpha=1 case through the
/// log-uniform inverse-CDF instead.
///
/// `sample()` returns ranks in `[0, k)`. Rank 0 is the most frequent.
///
/// Setup is O(k) (it sums `ζ(k, α) = Σ_{i=1..=k} 1/i^α`). For `k = 1 M` this
/// is ~10 ms — fine for the Pool=N use case where the pool builder runs once
/// at sender startup and setup time is captured in [`build_pool_timed`]'s
/// returned `Duration` (excluded from `wall_clock_ms`).
pub struct ZipfianSampler {
    rng: StdRng,
    alpha: f64,
    k: u64,
    /// `ζ(k, α)` — normalising constant for the truncated zeta distribution.
    zetan: f64,
    /// Quasi-uniform cutoff used by the inverse-CDF branch (alpha != 1).
    eta: f64,
    /// Pre-computed `0.5_f64.powf(alpha)` to avoid recomputing inside sample().
    half_alpha: f64,
}

impl ZipfianSampler {
    /// Construct a new sampler over ranks `[0, k)` with skew `alpha` and
    /// PRNG seed `seed`. Panics if `alpha <= 0.0` or `k == 0` — callers are
    /// expected to validate via `BlastShape::*` enums.
    pub fn new(alpha: f64, k: u64, seed: u64) -> Self {
        assert!(alpha > 0.0, "alpha must be > 0");
        assert!(k > 0, "k must be > 0");
        let zetan = zeta(k, alpha);
        let zeta2 = zeta(2, alpha);
        // For alpha == 1, the inverse-CDF formula divides by zero. We still
        // initialise eta to a finite value so debug builds don't trip a
        // NaN-flag; sample() detects alpha == 1 separately and uses a log-
        // uniform inverse instead.
        let eta = if (alpha - 1.0).abs() < f64::EPSILON {
            1.0
        } else {
            (1.0 - (2.0_f64 / k as f64).powf(1.0 - alpha)) / (1.0 - zeta2 / zetan)
        };
        let half_alpha = (0.5_f64).powf(alpha);
        Self {
            rng: StdRng::seed_from_u64(seed),
            alpha,
            k,
            zetan,
            eta,
            half_alpha,
        }
    }

    /// Sample one rank in `[0, k)`. Rank 0 is the most frequent.
    pub fn sample(&mut self) -> u64 {
        let u: f64 = self.rng.gen();
        let uz = u * self.zetan;
        if uz < 1.0 {
            return 0;
        }
        if uz < 1.0 + self.half_alpha {
            return 1;
        }
        // For alpha != 1 use Gray et al.'s inverse-CDF approximation; for
        // alpha == 1 (the most common fraud-shape default) use a log-uniform
        // inverse, which is exact for the harmonic-series limit.
        let v: f64 = self.rng.gen();
        let rank_f = if (self.alpha - 1.0).abs() < f64::EPSILON {
            // alpha == 1: cumulative ∝ ln(r); inverse is r = 2 * exp(v *
            // (ln(k) - ln(2))). Anchor at r=2 because we already special-
            // cased r=0 and r=1 above.
            let ln_k = (self.k as f64).ln();
            let ln_2 = 2.0_f64.ln();
            (2.0 * (v * (ln_k - ln_2)).exp()).floor()
        } else {
            (self.k as f64) * (self.eta * v - self.eta + 1.0).powf(1.0 / (1.0 - self.alpha))
        };
        let rank = rank_f as u64;
        rank.min(self.k - 1)
    }
}

/// Truncated zeta function `ζ(n, α) = Σ_{i=1..=n} 1 / i^α`.
fn zeta(n: u64, alpha: f64) -> f64 {
    let mut s = 0.0_f64;
    for i in 1..=n {
        s += 1.0 / (i as f64).powf(alpha);
    }
    s
}

/// Build N pre-encoded TCP frames matching the requested `BlastShape`.
///
/// Each entry is a complete `[u32 length][u16 op][u8 ct][payload]` frame —
/// the same envelope produced by `beava_core::wire::encode_frame` for
/// `OP_PUSH`. Callers can stream the pool out a TCP connection with
/// `write_all(&pool[i])` without touching the encoder again on the hot path.
pub fn build_pool(
    shape: BlastShape,
    cfg: &BlastShapeConfig,
    n: u64,
) -> Result<Vec<Bytes>, BlastShapeError> {
    match shape {
        BlastShape::Mixed { event_count } => {
            if cfg.event_names_for_mixed.len() < event_count {
                return Err(BlastShapeError::MixedRequiresMultipleEvents);
            }
        }
        BlastShape::Zipfian { alpha, cardinality } => {
            if alpha <= 0.0 {
                return Err(BlastShapeError::InvalidAlpha);
            }
            if cardinality == 0 {
                return Err(BlastShapeError::InvalidCardinality);
            }
        }
        BlastShape::Uniform { cardinality } => {
            if cardinality == 0 {
                return Err(BlastShapeError::InvalidCardinality);
            }
        }
        BlastShape::Fixed => {}
    }

    let mut rng = StdRng::seed_from_u64(cfg.seed);
    let mut zipf = if let BlastShape::Zipfian { alpha, cardinality } = shape {
        Some(ZipfianSampler::new(
            alpha,
            cardinality,
            cfg.seed.wrapping_add(0xDEAD),
        ))
    } else {
        None
    };

    let mut pool: Vec<Bytes> = Vec::with_capacity(n as usize);

    // Single reusable encode buffer to amortise allocation across the loop.
    let mut buf = BytesMut::with_capacity(4 * 1024);

    let ct = match cfg.wire_format {
        WireFormat::Json => CT_JSON,
        WireFormat::Msgpack => CT_MSGPACK,
    };

    // Fixed shape: encode the single frame once and clone its Bytes N times.
    // Both a perf-win (no per-frame encode) AND a contract guaranteed to the
    // tests (every entry byte-identical to entry 0).
    if let BlastShape::Fixed = shape {
        let body = build_event_body(
            cfg.pipeline,
            /* key_idx */ 0,
            /* seq */ 0,
            &mut rng,
        );
        let payload = encode_envelope(&cfg.pipeline.event_name, &body, cfg.wire_format);
        let frame = Frame {
            op: OP_PUSH,
            content_type: ct,
            payload: Bytes::from(payload),
        };
        buf.clear();
        encode_frame(&frame, &mut buf);
        let frozen: Bytes = buf.split().freeze();
        for _ in 0..n {
            pool.push(frozen.clone()); // Bytes::clone is a refcount bump
        }
        return Ok(pool);
    }

    for seq in 0..n {
        let key_idx: u64 = match shape {
            BlastShape::Fixed => unreachable!("handled above"),
            BlastShape::Uniform { cardinality } => rng.gen_range(0..cardinality),
            BlastShape::Zipfian { .. } => zipf.as_mut().unwrap().sample(),
            // Mixed varies the EVENT NAME, but key cardinality is the
            // pipeline's default (1 M); no shared key tracking needed.
            BlastShape::Mixed { .. } => rng.gen_range(0..1_000_000_u64),
        };
        let event_name: &str = match shape {
            BlastShape::Mixed { event_count } => {
                let idx = rng.gen_range(0..event_count);
                cfg.event_names_for_mixed[idx]
            }
            _ => &cfg.pipeline.event_name,
        };

        let body = build_event_body(cfg.pipeline, key_idx, seq, &mut rng);
        let payload = encode_envelope(event_name, &body, cfg.wire_format);

        let frame = Frame {
            op: OP_PUSH,
            content_type: ct,
            payload: Bytes::from(payload),
        };
        buf.clear();
        encode_frame(&frame, &mut buf);
        // `split` returns an owned BytesMut we can freeze without copying;
        // the remaining capacity stays on `buf` for the next iteration. No
        // extra allocation per frame.
        pool.push(buf.split().freeze());
    }

    Ok(pool)
}

/// Same as [`build_pool`], but additionally returns the wall-clock time spent
/// building the pool. The caller can subtract this from any `wall_clock_ms`
/// measurement (setup is excluded from saturation reads).
pub fn build_pool_timed(
    shape: BlastShape,
    cfg: &BlastShapeConfig,
    n: u64,
) -> Result<(Vec<Bytes>, Duration), BlastShapeError> {
    let t0 = Instant::now();
    let pool = build_pool(shape, cfg, n)?;
    Ok((pool, t0.elapsed()))
}

/// Build the per-event JSON `body` object that goes inside the envelope.
/// Field set matches `make_event_payload` in `beava-bench-v18.rs`:
/// `{key_field: "k%08u", event_time: i64, ...extra_fields}`.
fn build_event_body(pipeline: &PipelineConfig, key_idx: u64, seq: u64, rng: &mut StdRng) -> Value {
    let mut obj = serde_json::Map::with_capacity(2 + pipeline.extra_fields.len());
    obj.insert(
        pipeline.key_field.clone(),
        Value::String(format!("k{key_idx:08}")),
    );
    obj.insert(
        "event_time".to_string(),
        Value::Number((1_000_000 + seq as i64).into()),
    );
    for (field, ty) in &pipeline.extra_fields {
        let v = match ty.as_str().unwrap_or("f64") {
            "f64" => serde_json::json!(rng.gen_range(0.0..1000.0)),
            "i64" => serde_json::json!(rng.gen_range(0_i64..1_000_000)),
            "str" => serde_json::json!(format!("s{}", rng.gen_range(0..1000))),
            _ => serde_json::json!(0),
        };
        obj.insert(field.clone(), v);
    }
    Value::Object(obj)
}

/// Encode `{event, body}` into either JSON bytes or MessagePack bytes.
fn encode_envelope(event_name: &str, body: &Value, wire_format: WireFormat) -> Vec<u8> {
    match wire_format {
        WireFormat::Json => {
            let envelope = serde_json::json!({ "event": event_name, "body": body });
            serde_json::to_vec(&envelope).expect("json envelope encode")
        }
        WireFormat::Msgpack => {
            #[derive(Serialize)]
            struct Envelope<'a> {
                event: &'a str,
                body: &'a Value,
            }
            let env = Envelope {
                event: event_name,
                body,
            };
            rmp_serde::to_vec_named(&env).expect("msgpack envelope encode")
        }
    }
}
